File: src/main/java/br/com/jiratorio/mapper/UserConfigMapper.java
package br.com.jiratorio.mapper;

import br.com.jiratorio.domain.ImportHolidayInfo;
import br.com.jiratorio.domain.entity.UserConfig;
import br.com.jiratorio.domain.request.UpdateUserConfigRequest;
import br.com.jiratorio.domain.response.UserConfigResponse;
import br.com.jiratorio.mapper.transformer.CityTransformer;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.Mappings;

@Mapper(componentModel = "spring",
        uses = CityTransformer.class)
public interface UserConfigMapper {

    UserConfigResponse userConfigToResponse(UserConfig userConfig);

    @Mappings({ // @formatter:off
            @Mapping(target = "id",         ignore = true),
            @Mapping(target = "city",       qualifiedByName = "normalizeCity"),
            @Mapping(target = "updatedAt",  ignore = true),
            @Mapping(target = "createdAt",  ignore = true),
            @Mapping(target = "owner",      ignore = true),
            @Mapping(target = "lastEditor", ignore = true),
            @Mapping(target = "username",   ignore = true),
    }) // @formatter:on
    void updateFromRequest(@MappingTarget UserConfig userConfig, UpdateUserConfigRequest updateUserConfigRequest);

    default ImportHolidayInfo toImportHolidayInfo(UserConfig userConfig) {
        return new ImportHolidayInfo(userConfig.getState(), userConfig.getCity(), userConfig.getHolidayToken());
    }

}

File: src/main/java/br/com/jiratorio/mapper/UserConfigMapper.kt
package br.com.jiratorio.mapper

import br.com.jiratorio.domain.ImportHolidayInfo
import br.com.jiratorio.domain.entity.UserConfig
import br.com.jiratorio.domain.request.UpdateUserConfigRequest
import br.com.jiratorio.domain.response.UserConfigResponse
import br.com.jiratorio.extension.stripAccents
import org.springframework.stereotype.Component

@Component
class UserConfigMapper {

    fun userConfigToResponse(userConfig: UserConfig): UserConfigResponse {
        return UserConfigResponse(
            state = userConfig.state,
            city = userConfig.city,
            holidayToken = userConfig.holidayToken,
            leadTimeChartType = userConfig.leadTimeChartType.name,
            throughputChartType = userConfig.throughputChartType.name
        )
    }

    fun updateFromRequest(userConfig: UserConfig, updateUserConfigRequest: UpdateUserConfigRequest): UserConfig {
        return userConfig.apply {
            holidayToken = updateUserConfigRequest.holidayToken
            state = updateUserConfigRequest.state
            city = updateUserConfigRequest.city
                ?.stripAccents()
                ?.replace(" ", "")
                ?.toUpperCase()

            val leadTimeChartType = updateUserConfigRequest.leadTimeChartType
            if (leadTimeChartType != null) {
                this.leadTimeChartType = leadTimeChartType
            }

            val throughputChartType = updateUserConfigRequest.throughputChartType
            if (throughputChartType != null) {
                this.throughputChartType = throughputChartType
            }
        }
    }

    fun toImportHolidayInfo(userConfig: UserConfig): ImportHolidayInfo {
        return ImportHolidayInfo(
            state = userConfig.state!!,
            city = userConfig.city!!,
            holidayToken = userConfig.holidayToken!!
        )
    }

}
